query/aggregate.go

```diff
@@ -0,0 +1,224 @@
+	// First we delete all tags with names other than those found in 'tags'
+	newTags := api.NewTagSet()
+	for _, tag := range tags {
+		newTags[tag] = series.TagSet[tag]
+	}
+	// replace series' TagSet with newTags
+	series.TagSet = newTags
+
+	// Next, find the best bucket for this series:
+	for i, row := range rows {
+		if groupAccepts(row.TagSet, series.TagSet, tags) {
+			rows[i].List = append(rows[i].List, series)
+			return rows
+		}
+	}
+	// Otherwise, no bucket yet exists
+	return append(rows, group{
+		[]api.Timeseries{series},
+		newTags,
+	})
+}
+
+// Groups the given SeriesList by tags, producing a list of lists (of type groupResult)
+func groupBy(list api.SeriesList, tags []string) []group {
+	result := []group{}
+	for _, series := range list.Series {
+		result = addToGroup(result, series, tags)
+	}
+	return result
+}
+
+// The aggregator interface is the public-facing way in which values are aggregated.
+// Aggregator objects are required to perform aggregation (max, min, range, mean, sum, etc.)
+// Their interface is the `aggregate` method, which takes an array of floats.
+type aggregator interface {
+	aggregate([]float64) float64
+}
+
+type sumAggregator struct {
+}
+
+var _ aggregator = sumAggregator{}
+
+// The `aggregate()` for sum finds the sum of the given array.
+func (aggregator sumAggregator) aggregate(array []float64) float64 {
+	sum := 0.0
+	for _, v := range array {
+		sum += v
+	}
+	return sum
+}
+
+// A mean aggregator is highly similar to a sum aggregator.
+// It computes the aggregate mean for a seriesList
+type meanAggregator struct {
+}
+
+var _ aggregator = meanAggregator{}
+
+// The mean aggregator returns the mean of the given array
+func (aggregator meanAggregator) aggregate(array []float64) float64 {
+	if len(array) == 0 {
+		// The mean of an empty list is not well-defined
+		return math.NaN()
+	}
+	sum := 0.0
+	for _, v := range array {
+		sum += v
+	}
+	return sum / float64(len(array))
+}
+
+// The min aggregator is an aggregator that computes the aggregate minimum for a seriesList
+type minAggregator struct {
+}
+
+var _ aggregator = minAggregator{}
+
+func (aggregator minAggregator) aggregate(array []float64) float64 {
+	if len(array) == 0 {
+		// The minimum of an empty list is not well-defined
+		return math.NaN()
+	}
+	min := array[0]
+	for _, v := range array {
+		min = math.Min(min, v)
+	}
+	return min
+}
+
+// The maxAggregator is an aggregator that computes the aggregate maximum for a seriesList
+type maxAggregator struct {
+}
+
+var _ aggregator = maxAggregator{}
+
+func (aggregator maxAggregator) aggregate(array []float64) float64 {
+	if len(array) == 0 {
+		// The maximum of an empty list is not well-defined
+		return math.NaN()
+	}
+	max := array[0]
+	for _, v := range array {
+		max = math.Max(max, v)
+	}
+	return max
+}
+
+type aggregate int
+
+const (
+	sumAggregate aggregate = iota + 1
+	meanAggregate
+	minAggregate
+	maxAggregate
+)
+
+var aggregatorMap = map[aggregate]aggregator{
```

Comment:
Meant more of `map[aggregate]func([]float64)float64`. There's no need to go through a struct since these are all just different aggregate functions.